src/query/service/src/sql/optimizer/heuristic/where_optimizer.rs

```diff
@@ -0,0 +1,154 @@
+// Copyright 2022 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use common_exception::Result;
+
+use crate::sql::optimizer::ColumnSet;
+use crate::sql::optimizer::SExpr;
+use crate::sql::plans::Filter;
+use crate::sql::plans::LogicalGet;
+use crate::sql::plans::PatternPlan;
+use crate::sql::plans::Prewhere;
+use crate::sql::plans::RelOp;
+use crate::sql::plans::Scalar;
+use crate::sql::MetadataRef;
+
+pub struct WhereOptimizer {
+    metadata: MetadataRef,
+    pattern: SExpr,
+}
+
+impl WhereOptimizer {
+    pub fn new(metadata: MetadataRef) -> Self {
+        Self {
+            metadata,
+            pattern: SExpr::create_unary(
```

Comment:
can we merge this optimizer into `filter_pushdown`? The pattern is the same.

Reply:
LGTM

Reply:
> We should not merge it to `filter_pushdown`, because we need do this optimization after `prune_columns`. Or `prewhere.need_columns` will be wrong.

Could you please give an example?

Reply:
For example:

With table t2:

```
mysql> desc t2;
+-------+------+------+---------+-------+
| Field | Type | Null | Default | Extra |
+-------+------+------+---------+-------+
| a     | INT  | NO   | 0       |       |
| b     | INT  | NO   | 0       |       |
+-------+------+------+---------+-------+
```

`select a from t2 where a = 1;`

When matching the pattern for `rule_filter_scan_pushdown`, `LogicalGet.columns` is vec![0 ,1]:

![image](https://user-images.githubusercontent.com/33975039/187175587-952d3e63-8c51-4003-9b83-7d33895d8cd9.png)

After `prune_columns`, `LogicalGet.columns` will be changed to vec![0]:

![image](https://user-images.githubusercontent.com/33975039/187175796-61387d20-a945-4c13-bd46-c4c473366bcf.png)


If we do prewhere optimization within `rule_filter_scan_pushdown`, we pushed down prewhere information will be wrong.

Reply:
Ok. I get it.